### PR TITLE
[Merged by Bors] - feat(Algebra/Order/Archimedean): multiplicativize Archimedean

### PR DESCRIPTION
**Pull request by pechersky**
This allows one to talk about finding powers greater than some element for ordered monoids. Before, this was only available for ordered semirings. This is to later install the instance on valuation groups (`LinearOrderedCommGroupWithZero`). Install the instance on NNRat and NNReal (though they are ordered semirings). Also, add some missing instances about the nonnegative submonoid.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Order/Archimedean/Basic.lean [hunk 1]
```diff
@@ -38,74 +40,102 @@ class Archimedean (α) [OrderedAddCommMonoid α] : Prop where
...
 
-section LinearOrderedAddCommGroup
+@[to_additive]
+instance OrderDual.mulArchimedean [OrderedCommGroup α] [MulArchimedean α] : MulArchimedean αᵒᵈ :=
+  ⟨fun x y hy =>
+    let ⟨n, hn⟩ := MulArchimedean.arch (ofDual x)⁻¹ (inv_lt_one_iff_one_lt.2 hy)
+    ⟨n, by rwa [inv_pow, inv_le_inv_iff] at hn⟩⟩
 
-variable [LinearOrderedAddCommGroup α] [Archimedean α]
+instance Additive.Archimedean [OrderedCommGroup α] [MulArchimedean α] : Archimedean (Additive α) :=
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Order/Archimedean/Basic.lean:61]:
```suggestion
instance Additive.instArchimedean [OrderedCommGroup α] [MulArchimedean α] : Archimedean (Additive α) :=
```

File: Mathlib/Algebra/Order/Archimedean/Basic.lean [hunk 2]
```diff
@@ -38,74 +40,102 @@ class Archimedean (α) [OrderedAddCommMonoid α] : Prop where
...
 
-/-- An archimedean decidable linearly ordered `AddCommGroup` has a version of the floor: for
-`a > 0`, any `g` in the group lies between some two consecutive multiples of `a`. -/
-theorem existsUnique_zsmul_near_of_pos {a : α} (ha : 0 < a) (g : α) :
-    ∃! k : ℤ, k • a ≤ g ∧ g < (k + 1) • a := by
-  let s : Set ℤ := { n : ℤ | n • a ≤ g }
-  obtain ⟨k, hk : -g ≤ k • a⟩ := Archimedean.arch (-g) ha
-  have h_ne : s.Nonempty := ⟨-k, by simpa [s] using neg_le_neg hk⟩
-  obtain ⟨k, hk⟩ := Archimedean.arch g ha
+instance Multiplicative.MulArchimedean [OrderedAddCommGroup α] [Archimedean α] :
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Order/Archimedean/Basic.lean:64]:
```suggestion
instance Multiplicative.instMulArchimedean [OrderedAddCommGroup α] [Archimedean α] :
```

**Review by Ruben-VandeVelde — APPROVED** [review]

File: Mathlib/Algebra/Order/Archimedean/Basic.lean (hunk 1 above)
Comment (pechersky) [reply Mathlib/Algebra/Order/Archimedean/Basic.lean:61]: I can rename them as such -- should the existing `OrderDual.archimedean` be renamed also?

Comment (YaelDillies) [reply Mathlib/Algebra/Order/Archimedean/Basic.lean:61]: They should, yes

File: Mathlib/Algebra/Order/Archimedean/Basic.lean [hunk 3]
```diff
@@ -38,74 +40,102 @@ class Archimedean (α) [OrderedAddCommMonoid α] : Prop where
...
-instance OrderDual.archimedean [OrderedAddCommGroup α] [Archimedean α] : Archimedean αᵒᵈ :=
-  ⟨fun x y hy =>
-    let ⟨n, hn⟩ := Archimedean.arch (-ofDual x) (neg_pos.2 hy)
-    ⟨n, by rwa [neg_nsmul, neg_le_neg_iff] at hn⟩⟩
+section MulArchimedean
 
-variable {M : Type*}
+/-- An ordered commutative monoid is called `MulArchimedean` if for any two elements `x`, `y`
+such that `1 < y`, there exists a natural number `n` such that `x ≤ y ^ n`. -/
+class MulArchimedean (α) [OrderedCommMonoid α] : Prop where
```
Comment (YaelDillies) [inline Mathlib/Algebra/Order/Archimedean/Basic.lean:47]:
```suggestion
@[to_additive]
class MulArchimedean (α) [OrderedCommMonoid α] : Prop where
```

File: Mathlib/Algebra/Order/Nonneg/Ring.lean [hunk 4]
```diff
@@ -78,12 +78,24 @@ instance orderedCommSemiring [OrderedCommSemiring α] : OrderedCommSemiring { x
     (fun _ _ => rfl) (fun _ _ => rfl)
     (fun _ _ => rfl) (fun _ _ => rfl) fun _ => rfl
 
+instance orderedCommMonoid [OrderedCommSemiring α] : OrderedCommMonoid { x : α // 0 ≤ x } where
+  mul_le_mul_left a _ h c := mul_le_mul le_rfl h a.prop c.prop
```
Comment (YaelDillies) [inline Mathlib/Algebra/Order/Nonneg/Ring.lean:82]: Why do you need this? We already have `Nonneg.orderedCommSemiring`

Comment (pechersky) [reply Mathlib/Algebra/Order/Nonneg/Ring.lean]: How does that help? There is no `OrderedCommSemiring.toOrderedCommMonoid`. It's only true for the `Nonneg` cone.

Comment (YaelDillies) [reply Mathlib/Algebra/Order/Nonneg/Ring.lean:82]: Oh right!

File: Mathlib/Algebra/Order/Nonneg/Ring.lean
Comment (YaelDillies) [inline Mathlib/Algebra/Order/Nonneg/Ring.lean:1]: Strictly speaking this is out of scope

File: Mathlib/Data/NNReal/Basic.lean
Comment (YaelDillies) [inline Mathlib/Data/NNReal/Basic.lean:1]: and this too

**Review by YaelDillies — COMMENTED** [review]
... apart from that,

maintainer merge

File: Mathlib/Algebra/Order/Nonneg/Ring.lean
Comment (pechersky) [reply Mathlib/Algebra/Order/Nonneg/Ring.lean:1]: This one was necessary for `Nonneg.instMulArchimedean`

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**